core/workflow_settings.ts

```diff
@@ -54,6 +55,18 @@ function verifyWorkflowSettingsAsJson(workflowSettingsAsJson: object): dataform.
     // tslint:disable-next-line: no-string-literal
     projectConfig.warehouse = "bigquery";
   }
+
+  if (!projectConfig.version) {
```

Comment:
I'm not really following this.

If the project is manually installing @dataform/core, I don't think we should set this field with a default. The version being undefined in the projectConfig is actually useful to know!

The validation check - also not sure why we need this. This would only happen in the case that the version specified in workflow_settings was not automatically installed by the CLI or Dataform on GCP. Neither should be possible!

Reply:
Can you elaborate on what the version being undefined in the projectConfig useful for? I think all cases are covered by the version not being exported by core.

For the validation check, if neither should be possible for either the CLI or Dataform on GCP, then why not check this here!

The Dataform Core interface currently allows all of the `package.json`, the `workflow_settings.yaml` file, and the currently installed Dataform Core version under `node_modules` to be different. By enforcing an opinion on where the version should actually come from, it makes the UX waaay simpler, and prevents every Dataform consumer from having its own implementation.

An example UX:
* Dataform user has an Airflow configuration, using the new `workflow_settings.yaml`.
* They check in a package.json, because they want core to be installed via `npm install`.
* Someone bumps the `workflow_settings.yaml`, but it has no effect.
* With this CL, it will be clear that just bumping the version in `workflow_settings.yaml` does nothing.

Reply:
Also if someone has a `package.json` with multiple other packages specified, then this works well with `workflow_settings.yaml` automatically interpreting the Dataform Version installed.

Reply:
ProjectConfig is a proto representation of workflow_settings.yaml or dataform.json. But you're using it to track useful info about the compilation, but we already track the version of dataform/core used for compilation here: https://github.com/dataform-co/dataform/blob/main/protos/core.proto#L296

So I think in general we should not be writing to `ProjectConfig` except to define static default values, which is all we to it before this CL as far as I'm aware.

In terms of knowing if it's undefined or not - this is kind of useful as it tells you where the version came from - was it actually defined in workflow_settings and therefore installed automatically, or, did we use a user installed version of dataform/core that already existed inside node_modules (probably from package.json)

I'm OK with the equality check 👍

Reply:
> Someone bumps the workflow_settings.yaml, but it has no effect.

This should never be the case, as we are always going to give `version` in workflow settings precedence right? Or if there is a conflict and both are specified, we will throw an error. Either way it will not be ignored.

Reply:
I agree with Lewis.

We're never going to _ignore_ the case where the version is defined both in workflow settings and package.json (I would personally argue for throwing an error, as this seems like a very easy path to lead to user confusion otherwise!).

We shouldn't be treating `ProjectConfig` as an _output_.

Reply:
Ok I think I can buy not setting it with the current version if unset. The exact UX for this will become more clear with the internal changes and tests for them.

It sounds like there's agreement now that this is a good place for the validation check? `projectConfig.dataformCoreVersion !== version`